src/main/java/org/opentripplanner/routing/api/request/request/filter/SelectRequest.java

```diff
@@ -0,0 +1,158 @@
+package org.opentripplanner.routing.api.request.request.filter;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.opentripplanner.framework.tostring.ToStringBuilder;
+import org.opentripplanner.model.modes.AllowTransitModeFilter;
+import org.opentripplanner.routing.core.RouteMatcher;
+import org.opentripplanner.transit.model.basic.MainAndSubMode;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.network.Route;
+import org.opentripplanner.transit.model.timetable.TripTimes;
+
+public class SelectRequest implements Serializable {
+
+  public static class Builder {
+
+    private List<MainAndSubMode> transportModes = new ArrayList<>();
+    private List<FeedScopedId> agencies = new ArrayList<>();
+    private RouteMatcher routes = RouteMatcher.emptyMatcher();
+
+    // TODO: 2022-11-29 filters: group of routes
+
+    public Builder withTransportModes(List<MainAndSubMode> transportModes) {
+      this.transportModes = transportModes;
+
+      return this;
+    }
+
+    public Builder addTransportMode(MainAndSubMode transportMode) {
+      this.transportModes.add(transportMode);
+
+      return this;
+    }
+
+    public Builder withAgenciesFromString(String s) {
+      if (!s.isEmpty()) {
+        this.agencies = FeedScopedId.parseListOfIds(s);
+      }
+
+      return this;
+    }
+
+    public Builder withAgencies(List<FeedScopedId> agencies) {
+      this.agencies = agencies;
+
+      return this;
+    }
+
+    public Builder withRoutesFromString(String s) {
+      if (!s.isEmpty()) {
+        this.routes = RouteMatcher.parse(s);
+      } else {
+        this.routes = RouteMatcher.emptyMatcher();
+      }
+
+      return this;
+    }
+
+    public Builder withRoutes(RouteMatcher routes) {
+      this.routes = routes;
+
+      return this;
+    }
+
+    public SelectRequest build() {
+      return new SelectRequest(this);
+    }
+  }
+
+  public static Builder of() {
+    return new Builder();
+  }
+
+  public SelectRequest(Builder builder) {
+    if (!builder.transportModes.isEmpty()) {
+      this.transportModeFilter = AllowTransitModeFilter.of(builder.transportModes);
+    } else {
+      this.transportModeFilter = null;
+    }
+
+    // TODO: 2022-12-20 filters: having list of modes and modes filter in same instance is not very elegant
+    this.transportModes = builder.transportModes;
```

Comment:
Can we not combine these somehow?

Reply:
Would be nice. The problem is that we need them for toString and it is not so easy to get this information from AllowTransitModeFilter since it is an interface with multiple implementations. Previously we were creating filter on the spot but with the new implementation that would be a lot of instances. It's better to create them ahead of time. unfortunately this means that we have to keep transportModes input stored somewhere so that we can generate toString info when needed.